src/plugins/visualizations/public/utils/saved_visualize_utils.ts

```diff
@@ -0,0 +1,392 @@
+    isTitleDuplicateConfirmed = false,
+    onTitleDuplicate,
+  }: {
+    confirmOverwrite?: boolean;
+    isTitleDuplicateConfirmed?: boolean;
+    onTitleDuplicate?: () => void;
+  },
+  services: {
+    savedObjectsClient: SavedObjectsClientContract;
+    chrome: ChromeStart;
+    overlays: OverlayStart;
+    savedObjectsTagging?: SavedObjectsTaggingApi;
+  }
+) {
+  // Save the original id in case the save fails.
+  const originalId = savedObject.id;
+  // Read https://github.com/elastic/kibana/issues/9056 and
+  // https://github.com/elastic/kibana/issues/9012 for some background into why this copyOnSave variable
+  // exists.
+  // The goal is to move towards a better rename flow, but since our users have been conditioned
+  // to expect a 'save as' flow during a rename, we are keeping the logic the same until a better
+  // UI/UX can be worked out.
+  if (savedObject.copyOnSave) {
+    delete savedObject.id;
+  }
+
+  const attributes: any = {
+    visState: JSON.stringify(savedObject.visState),
+    title: savedObject.title,
+    uiStateJSON: savedObject.uiStateJSON,
+    description: savedObject.description,
+    savedSearchId: savedObject.savedSearchId,
+    version: savedObject.version,
+  };
+  let references: any = [];
+
+  if (savedObject.searchSource) {
+    const { searchSourceJSON, references: searchSourceReferences } =
+      savedObject.searchSource.serialize();
+    attributes.kibanaSavedObjectMeta = { searchSourceJSON };
+    references.push(...searchSourceReferences);
+  }
+
+  if (savedObject.searchSourceFields) {
+    const [searchSourceFields, searchSourceReferences] = extractSearchSourceReferences(
+      savedObject.searchSourceFields
+    );
+    const searchSourceJSON = JSON.stringify(searchSourceFields);
+    attributes.kibanaSavedObjectMeta = { searchSourceJSON };
+    references.push(...searchSourceReferences);
+  }
+
+  if (services.savedObjectsTagging) {
+    references = services.savedObjectsTagging.ui.updateTagsReferences(
+      references,
+      savedObject.tags || []
+    );
+  }
+
+  const extractedRefs = extractReferences({ attributes, references });
+
+  if (!extractedRefs.references) {
+    throw new Error('References not returned from extractReferences');
+  }
+
+  try {
+    await checkForDuplicateTitle(
+      savedObject as any,
+      isTitleDuplicateConfirmed,
+      onTitleDuplicate,
+      services as any
+    );
+    savedObject.isSaving = true;
+
+    const createOpt = {
+      id: savedObject.id,
+      migrationVersion: savedObject.migrationVersion,
+      references: extractedRefs.references,
+    };
+    const resp = confirmOverwrite
+      ? await saveWithConfirmation(attributes, savedObject, createOpt, services)
+      : await services.savedObjectsClient.create(SAVED_VIS_TYPE, extractedRefs.attributes, {
+          ...createOpt,
+          overwrite: true,
+        });
+
+    services.chrome.recentlyAccessed.add(getFullPath(resp.id), savedObject.title, String(resp.id));
```

Comment:
I prefer when we write simpler functions who are not responsible for many actions, it seems to me that this keys should be processed outside the `save` function, as an example, see Discover

Reply:
moved directly in places where we call save